activerecord/lib/active_record/attribute_methods/write.rb

```diff
@@ -36,8 +36,7 @@ def write_attribute(attr_name, value)
         end
 
         name = self.class.primary_key if name == "id".freeze && self.class.primary_key
```

Comment:
You read my mind (or my `git stash`)!

It won't be possible until I revive https://github.com/rails/rails/pull/28395 so that `sync_with_transaction_state` is called when accessing a custom primary key. Doing that is blocked on this PR, because I want to use `_write_attribute` in the implementation.

Another thing I want to try is moving the attribute alias checks to a separate module that's only included in the model when the first alias is added - essentially a deopt for the unlikely branch, as most models (in my experience) don't have any aliases.